pkg/controllers/node/emptiness.go

```diff
@@ -42,6 +42,14 @@ func (r *Emptiness) Reconcile(ctx context.Context, provisioner *v1alpha5.Provisi
 	if provisioner.Spec.TTLSecondsAfterEmpty == nil {
 		return reconcile.Result{}, nil
 	}
+
+	// If the node just launched, give it some time to fully initialize and for kube-scheduler to assign pods to the
+	// node. This also allows for extended resource device plugins to initialize and assign resource capacity/allocatable
+	// quantities to the node status.
+	if n.CreationTimestamp.After(injectabletime.Now().Add(-5 * time.Minute)) {
```

Comment:
I don't feel great about this, but I'm not sure of a better solution.  The nvidia driver plugin DS takes some amount of time to start and for the GPU resources to be recorded on the nodes capacity/allocatable.  In my testing this was greater than thirty seconds so the nodes were being deleted before the driver ever finished initializing.

Reply:
Could we check that all the extended resource annotations have been added to the node capacity & allocatable as a trigger for when to remove the `karpenter.sh/not-ready` annotation?

Reply:
o dang :(   I guess you could still replace the time based check and instead check for the annotated resources to be populated in resources.

Reply:
The one positive side effect of the time based change is that we don't see spurious empty TTL annotations getting added and removed which happens every time as we notice the empty node after it's ready before kube-scheduler binds pods to it.

Reply:
There are two time based checks that have been added in this PR, one is used for emptiness and the other for computing allocatable resources.

They are both working around the same fundamental issue, a node becomes "Ready" and then it takes some amount of time for device plugins to initialize and register the resources on the node so that pods can be scheduled to it.

The emptiness one is needed as if the node starts up and we call it "empty" as soon as it's ready, it may remain empty and be terminated before the device plugin can finish initializing.

The other is used when determining the remaining allocatable resources of an in-flight node. 

One alternative is to increase the time to 15 minutes, but only use it on nodes with extended resources?

Reply:
Could we use karpenter's readiness taint as a signal and only remove it once all daemonsets are passing readiness probes?

Reply:
The Nvidia plugin daemonset doesn't tolerate our readiness taint, so it doesn't get scheduled until that taint is removed.

Reply:
I'm really wary of time based solutions, since this varies broadly across vendors and is one of the key pain points of cluster autoscaler that karpenter has managed to avoid so far. I'm open minded to an initial impl that uses timeouts, but I want to find a long term path that fundamentally solves the pod coordination problem.

Reply:
I don't like it as well, but I haven't come up with a better solution yet. Maybe eliminating the time based readiness gate and replacing it with an 'all expected daemonsets are ready' gate would work?  We would need to ensure that all of the extended resource daemonsets that we support don't report 'ready' until they've been able to register the resources on the node:
- aws-vpc-cni
- Nvidia GPU
- AMD GPU
- AWS Neuron